### PR TITLE
let unnecessary_cast work for trivial non_literal expressions

### LINKED ISSUE #9562: clippy::unnecessary_cast does not warn on basic usage
### Summary

When casting the result of a function to the same type, `clippy::unnecessary_cast` doesn't warn

### Lint Name

clippy::unnecessary_cast

### Reproducer

I tried this code:

```rust
fn foo() -> f32 { 0. }

fn main() {
    let num = foo() as f32;
}
```

I expected to see this happen: Warning that the cast is unnecessary

Instead, this happened: No warning

### Version

_No response_

### PR DESCRIPTION
**Pull request by TennyZhuang**
Signed-off-by: TennyZhuang <zty0826@gmail.com>

---

changelog: [`unnecessary_cast`]: fix for trivial non_literal expressions

Fixes #9562

### CODE REVIEW

**Comment by TennyZhuang** [comment]
Some other tests were broken, convert to draft temporally :(

**Comment by TennyZhuang** [comment]
I've added some rules on some test files, but I'm not sure why they doesn't work.

 [783ee63](https://github.com/rust-lang/rust-clippy/pull/9576/commits/783ee635559a522a7a46b95419ccc536c6a28f97)

**Comment by kraktus** [comment]
Those tests need to be blessed (because we also check the code output after `clippy --fix`, which is stored as `my_test.fixed`.

You can bless them with `cargo dev bless` and commit the changes (which normally should just be adding the `allow` parameter to the corresponding `.fixed`)

File: clippy_lints/src/casts/unnecessary_cast.rs [hunk 1]
```diff
@@ -30,8 +30,10 @@ pub(super) fn check<'tcx>(
         }
     }
 
+    let cast_str = snippet_opt(cx, cast_expr.span).unwrap_or_default();
```
Comment (kraktus) [inline clippy_lints/src/casts/unnecessary_cast.rs:33]: nit, use [clippy_utils](https://doc.rust-lang.org/nightly/nightly-rustc/clippy_utils/index.html)::[source](https://doc.rust-lang.org/nightly/nightly-rustc/clippy_utils/source/index.html)::[snippet](https://doc.rust-lang.org/nightly/nightly-rustc/clippy_utils/source/fn.snippet.html#) instead (with canonical fallback being `..`, as shown in examples

File: clippy_lints/src/casts/unnecessary_cast.rs [hunk 2]
```diff
@@ -80,6 +82,17 @@ pub(super) fn check<'tcx>(
...
+        span_lint_and_sugg(
+            cx,
+            UNNECESSARY_CAST,
+            expr.span,
+            &format!("casting to the same type is unnecessary (`{cast_from}` -> `{cast_to}`)"),
+            "try",
+            cast_str,
+            Applicability::MachineApplicable,
+        );
+        return true;
```
Comment (kraktus) [inline clippy_lints/src/casts/unnecessary_cast.rs:95]: Looks like it should already be caught by https://github.com/rust-lang/rust-clippy/blob/783ee635559a522a7a46b95419ccc536c6a28f97/clippy_lints/src/casts/unnecessary_cast.rs#L71
no?

**Comment by TennyZhuang** [comment]
@kraktus Hi, in fact, I blessed them using `cargo dev bless` but found the `allow` does not work. You can run them locally or check [the CI log](https://github.com/rust-lang/rust-clippy/actions/runs/3168136112/jobs/5159119802#step:7:976). The warnings are still produced so I don't push them.

I found that if I use the function level `#[allow(clippy::unnecessary_cast)]`, then everything works. I guess there are another bug but I will use it as a workaround.

File: clippy_lints/src/casts/unnecessary_cast.rs (hunk 2 above)
Comment (TennyZhuang) [reply clippy_lints/src/casts/unnecessary_cast.rs:95]: It's under a condition so it will only work for numeric literal.

Comment (kraktus) [reply clippy_lints/src/casts/unnecessary_cast.rs:95]: right, you could refactor it by turning your `if else` into a `if` and remove the afore mentioned line then I think

**Comment by TennyZhuang** [comment]
Sorry, I misunderstood and forget to run `cargo uitest` before `cargo dev bless` :(

**Comment by kraktus** [comment]
No worries!

File: clippy_lints/src/casts/unnecessary_cast.rs (hunk 2 above)
Comment (TennyZhuang) [reply clippy_lints/src/casts/unnecessary_cast.rs:95]: I've tried it when I implement it, but the control flow is too complicated to do that. At least, I will try to extract a common function `lint_same_type`.

Comment (TennyZhuang) [reply clippy_lints/src/casts/unnecessary_cast.rs]: Updated

File: clippy_lints/src/casts/unnecessary_cast.rs (hunk 1 above)
Comment (TennyZhuang) [reply clippy_lints/src/casts/unnecessary_cast.rs:33]: The original code uses `snippet_opt,` and the only thing I do is move the variable; there are also some other `snippet_opt` usages in the file, so I'm not sure if I should update them all in the PR. It seems a little off-topic.

**Comment by llogiq** [comment]
Looks good to me.

Thank you!

@bors r+

**Pull request merged**